File: python/locus/_worktrees.py
"""Managed Git worktrees and reusable, private Unity project slots."""
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any, TYPE_CHECKING

from ._models import WorkspaceRef
from ._scope import workspace_payload

if TYPE_CHECKING:
    from ._client import Client


@dataclass(frozen=True, slots=True)
class Worktree:
    checkout_id: str
    project_id: str
    root: str
    repo_root: str
    project_relative_path: str
    branch: str | None
    head_oid: str
    materialization_epoch: int
    managed: bool
    lifecycle: str
    dirty: bool
    pool_slot: bool
    assignment_id: str | None
    editor_version: str | None
    last_error: str | None
    workspace_ref: WorkspaceRef

    @classmethod
    def from_payload(cls, row: dict[str, Any]) -> Worktree:
        return cls(checkout_id=row["checkoutId"], project_id=row["projectId"],
            root=row["root"], repo_root=row["repoRoot"], project_relative_path=row["projectRelativePath"],
            branch=row.get("branch"), head_oid=row["headOid"], materialization_epoch=row["materializationEpoch"],
            managed=row["managed"], lifecycle=row["lifecycle"], dirty=row["dirty"], pool_slot=row["poolSlot"],
            assignment_id=row.get("assignmentId"), editor_version=row.get("editorVersion"),
            last_error=row.get("lastError"), workspace_ref=WorkspaceRef.from_payload(row["workspaceRef"]))


@dataclass(frozen=True, slots=True)
class PoolAcquisition:
    worktree: Worktree
    reused: bool
    preserved_library: bool


class Worktrees:
    def __init__(self, client: Client) -> None:
        self._client = client

    async def _call(self, action: str, params: dict[str, Any] | None = None, *, workspace_ref: Any = None) -> Any:
        return await self._client.rpc("worktrees." + action, {
            "workspaceRef": workspace_payload(workspace_ref),
            "executionDelegation": os.environ.get("LOCUS_SDK_EXECUTION_DELEGATION"),
            **(params or {}),
        }, timeout=300)

    async def list(self, *, workspace_ref: Any = None) -> list[Worktree]:
        """List recorded siblings, including idle/removed slots, without starting services."""
        return [Worktree.from_payload(row) for row in await self._call("list", workspace_ref=workspace_ref)]

    async def get(self, checkout_id: str, *, workspace_ref: Any = None) -> Worktree:
        """Explicitly resolve a current sibling and register its active runtime."""
        if not checkout_id.strip():
            raise ValueError("checkout_id cannot be empty")
        return Worktree.from_payload(await self._call("get", {"checkoutId": checkout_id.strip()}, workspace_ref=workspace_ref))

    async def create(self, *, destination: str, branch: str, start_ref: str | None = None,
        include_dirty: bool = False, workspace_ref: Any = None) -> Worktree:
        if not destination.strip() or not branch.strip():
            raise ValueError("destination and branch cannot be empty")
        return Worktree.from_payload(await self._call("create", {
            "destination": destination, "branch": branch, "startRef": start_ref, "includeDirty": include_dirty,
        }, workspace_ref=workspace_ref))

    async def discover(self, *, workspace_ref: Any = None) -> list[str]:
        return await self._call("discover", workspace_ref=workspace_ref)

    async def import_worktree(self, target_root: str, *, workspace_ref: Any = None) -> Worktree:
        if not target_root.strip():
            raise ValueError("target_root cannot be empty")
        return Worktree.from_payload(await self._call("import", {"targetRoot": target_root}, workspace_ref=workspace_ref))

    async def remove(self, worktree: Worktree | WorkspaceRef, *, workspace_ref: Any = None) -> None:
        """Remove a clean, closed, unassigned managed checkout; never force-delete user work."""
        await self._call("remove", {"worktree": workspace_payload(worktree)}, workspace_ref=workspace_ref)

    async def operations(self, *, workspace_ref: Any = None) -> list[dict[str, Any]]:
        return await self._call("operations", workspace_ref=workspace_ref)

    async def acquire(self, *, pool_root: str, commit: str, max_slots: int,
        branch: str | None = None, assignment_id: str | None = None, workspace_ref: Any = None) -> PoolAcquisition:
        """Assign a private Unity slot at an exact commit; reuse compatible idle Library."""
        if not pool_root.strip() or not commit.strip():
            raise ValueError("pool_root and commit cannot be empty")
        if isinstance(max_slots, bool) or not isinstance(max_slots, int) or max_slots <= 0:
            raise ValueError("max_slots must be a positive integer")
        row = await self._call("acquire", {"poolRoot": pool_root, "commit": commit,
            "maxSlots": max_slots, "branch": branch, "assignmentId": assignment_id}, workspace_ref=workspace_ref)
        return PoolAcquisition(Worktree.from_payload(row["worktree"]), row["reused"], row["preservedLibrary"])

    async def release(self, worktree: Worktree, *, workspace_ref: Any = None) -> Worktree:
        """Return the exact assignment after its editor is closed and checkout is clean."""
        if not isinstance(worktree, Worktree) or not worktree.assignment_id:
            raise ValueError("release requires the assigned Worktree returned by acquire")
        return Worktree.from_payload(await self._call("release", {
            "worktree": workspace_payload(worktree), "assignmentId": worktree.assignment_id,
        }, workspace_ref=workspace_ref))

File: src-tauri/src/sdk/worktrees.rs
//! SDK adapters use the same journal, registration and retirement as the UI.
use super::*;
use crate::workspace_service::{worktrees as store, CheckoutId, ProjectRegistry, WorkspaceRef};

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct Params {
    workspace_ref: WorkspaceRef,
    execution_delegation: Option<String>,
    worktree: Option<WorkspaceRef>,
    checkout_id: Option<String>,
    destination: Option<String>,
    branch: Option<String>,
    start_ref: Option<String>,
    #[serde(default)]
    include_dirty: bool,
    target_root: Option<String>,
    pool_root: Option<String>,
    commit: Option<String>,
    max_slots: Option<usize>,
    assignment_id: Option<String>,
}

fn required(value: Option<String>, name: &str) -> Result<String, String> {
    value
        .filter(|v| !v.trim().is_empty())
        .ok_or_else(|| format!("{name} is required"))
}

fn descriptor(registry: &ProjectRegistry, record: store::ManagedWorktree) -> Result<Value, String> {
    let id = CheckoutId::new(&record.checkout_id).map_err(|e| e.to_string())?;
    let generation = registry
        .runtime(&id)
        .filter(|runtime| runtime.materialization_epoch() == record.materialization_epoch)
        .map(|runtime| runtime.generation());
    let reference = WorkspaceRef::new(id, generation)
        .with_materialization_epoch(Some(record.materialization_epoch));
    let mut value = serde_json::to_value(record).map_err(|e| e.to_string())?;
    value["workspaceRef"] = serde_json::to_value(reference).map_err(|e| e.to_string())?;
    Ok(value)
}

fn target_epoch(reference: &WorkspaceRef) -> Result<u64, String> {
    reference.expected_materialization_epoch.ok_or_else(||
        "Worktree retirement requires its explicit materialization epoch; query worktrees.get first".into())
}

pub(super) async fn dispatch(app: &AppHandle, action: &str, value: Value) -> Result<Value, String> {
    use crate::commands;
    let params: Params = parse_params(value)?;
    let scope = resolve_sdk_workspace_scope(app, &params.workspace_ref, "worktrees")?;
    let source = scope.runtime().root().to_string_lossy().into_owned();
    let registry = app.state::<Arc<ProjectRegistry>>();
    let sessions = app.state::<Arc<SessionStore>>();
    match action {
        "list" => {
            let records = commands::list_managed_worktrees(source, registry.clone()).await?;
            records
                .into_iter()
                .map(|row| descriptor(&registry, row))
                .collect::<Result<Vec<_>, _>>()
                .map(Value::Array)
        }
        "get" => {
            let id = required(params.checkout_id, "checkoutId")?;
            let record = commands::list_managed_worktrees(source, registry.clone())
                .await?
                .into_iter()
                .find(|row| row.checkout_id == id)
                .ok_or("Checkout is outside the source project/repository")?;
            if record.lifecycle == "active" {
                // An explicit get can reopen the runtime; it never accepts a stale handle.
                commands::register_record(&registry, &sessions, &record)?;
            }
            descriptor(&registry, record)
        }
        "discover" => serde_json::to_value(commands::discover_worktrees(source, registry).await?)
            .map_err(|e| e.to_string()),
        "operations" => {
            serde_json::to_value(commands::get_worktree_operations(source, registry).await?)
                .map_err(|e| e.to_string())
        }
        "create" | "import" | "acquire" => {
            // Ordinary materialization reads an immutable commit and is already
            // coordinated by the worktree journal/slot lease. Only an undo-enabled
            // dirty snapshot participates in the caller's existing write gate.
            let _guard = if action == "create"
                && params.include_dirty
                && app
                    .state::<Arc<crate::config::AppConfig>>()
                    .session_undo_enabled()
            {
                Some(
                    crate::merge_jobs::coordination::acquire(
                        app,
                        scope.runtime(),
                        "worktrees.create",
                        params.execution_delegation.as_deref(),
                    )
                    .await?,
                )
            } else {
                None
            };
            if action == "create" {
                let record = commands::create_worktree(
                    store::CreateWorktreeRequest {
                        source_root: source,
                        destination: required(params.destination, "destination")?,
                        branch: required(params.branch, "branch")?,
                        start_ref: params.start_ref,
                        include_dirty: params.include_dirty,
                    },
                    registry.clone(),
                    sessions,
                )
                .await?;
                return descriptor(&registry, record);
            }
            if action == "import" {
                let record = commands::import_worktree(
                    source,
                    required(params.target_root, "targetRoot")?,
                    registry.clone(),
                    sessions,
                )
                .await?;
                return descriptor(&registry, record);
            }
            let acquired = commands::acquire_unity_project_slot(
                crate::workspace_service::pool::AcquirePoolRequest {
                    source_root: source,
                    pool_root: required(params.pool_root, "poolRoot")?,
                    commit: required(params.commit, "commit")?,
                    branch: params.branch,
                    max_slots: params
                        .max_slots
                        .filter(|v| *v > 0)
                        .ok_or("maxSlots must be positive")?,
                    assignment_id: params.assignment_id,
                },
                registry.clone(),
                sessions,
                Some(true),
            )
            .await?;
            Ok(
                json!({"worktree": descriptor(&registry, acquired.worktree)?,
                "reused": acquired.reused, "preservedLibrary": acquired.preserved_library}),
            )
        }
        "remove" | "release" => {
            let target = params.worktree.ok_or("worktree reference is required")?;
            let epoch = target_epoch(&target)?;
            if target.checkout_id == *scope.runtime().checkout_id() {
                return Err(
                    "Cannot retire the SDK's source checkout; use a sibling as workspace_ref"
                        .into(),
                );
            }
            // Do not take a target RunningTask lease here: retirement must see no users.
            // Commands recheck repository ownership, epoch, dirty state and Editor closure.
            if let Some(generation) = target.expected_generation {
                if registry
                    .runtime(&target.checkout_id)
                    .is_some_and(|r| r.generation() != generation)
                {
                    return Err(
                        "Worktree runtime generation is stale; query worktrees.get first".into(),
                    );
                }
            }
            if action == "remove" {
                commands::remove_managed_worktree(
                    source,
                    target.checkout_id.to_string(),
                    epoch,
                    registry,
                )
                .await?;
                return Ok(Value::Null);
            }
            let record = commands::release_unity_project_slot(
                source,
                target.checkout_id.to_string(),
                required(params.assignment_id, "assignmentId")?,
                epoch,
                registry.clone(),
            )
            .await?;
            descriptor(&registry, record)
        }
        _ => Err(format!("Unknown worktrees action '{action}'")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn retirement_requires_explicit_epoch_and_requests_reject_typos() {
        let reference = WorkspaceRef::new(CheckoutId::new("checkout-a").unwrap(), None);
        assert!(target_epoch(&reference).is_err());
        assert_eq!(
            target_epoch(&reference.with_materialization_epoch(Some(3))).unwrap(),
            3
        );
        assert!(serde_json::from_value::<Params>(json!({
            "workspaceRef": {"checkoutId":"source"}, "includeDity": true,
        }))
        .is_err());
    }
}
